Add option none to zfs redundant_metadata property

akashb-22 opened the pull request:
Currently, additional/extra copies are created for metadata in
addition to the redundancy provided by the pool(mirror/raidz/draid),
due to this 2 times more space is utilized per inode and this decreases
the total number of inodes that can be created in the filesystem. By
setting redundant_metadata to none, no additional copies of metadata
are created, hence can reduce the space consumed by the additional
metadata copies and increase the total number of inodes that can be
created in the filesystem.

Reviewed-by: Dipak Ghosh <dipak.ghosh@hpe.com>
Signed-off-by: Akash B <akash-b@hpe.com>

<!--- Please fill out the following template, which will help other contributors review your Pull Request. -->

<!--- Provide a general summary of your changes in the Title above -->

<!---
Documentation on ZFS Buildbot options can be found at
https://openzfs.github.io/openzfs-docs/Developer%20Resources/Buildbot%20Options.html
-->

### Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
By default, zfs takes extra copies of the metadata which leads to additional space usage and a total reduction in the number of inodes in a filesystem. 
<!--- If it fixes an open issue, please link to the issue here. -->

### Description
<!--- Describe your changes in detail -->
This patch adds the new option "none" to zfs redundant_metadata property which can be used to disable the additional metadata copies.

### How Has This Been Tested?
We have tested the above patch on VMs and on servers(HDD) along with lustre. 
Tested this patch by:
Creating 25 Million directories when redundant_metadata is set to "all" and "none". 
when,
redundant_metadata is set to "all" space USED is `591.3G` in `1368.64 ops/sec`
redundant_metadata is set to "none" space USED is `296.4G` in `2966.02 ops/sec`

Creating 25 Million 0k files when redundant_metadata is set to "all" and "none". 
when,
redundant_metadata is set to "all" space USED is `20.1G` in `10344.04 ops/sec`
redundant_metadata is set to "none" space USED is `10.3G` in `22050.56 ops/sec`

<!--- Please describe in detail how you tested your changes. --> 
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
<!--- If your change is a performance enhancement, please provide benchmarks here. -->
<!--- Please think about using the draft PR feature if appropriate -->

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Library ABI change (libzfs, libzfs\_core, libnvpair, libuutil and libzfsbootenv)
- [x] Documentation (a change to man pages or other documentation)

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the OpenZFS [code style requirements](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/openzfs/zfs/tree/master/tests) to cover my changes.
- [x] I have run the ZFS Test Suite with this change applied.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

freqlabs commented on module/zfs/dmu.c:
```diff
@@ -1989,11 +1989,13 @@ dmu_write_policy(objset_t *os, dnode_t *dn, int level, int wp, zio_prop_t *zp)
 
-		if (os->os_redundant_metadata == ZFS_REDUNDANT_METADATA_ALL ||
+		if ((os->os_redundant_metadata !=
+		    ZFS_REDUNDANT_METADATA_NONE &&
+		    (os->os_redundant_metadata == ZFS_REDUNDANT_METADATA_ALL ||
 		    (os->os_redundant_metadata ==
 		    ZFS_REDUNDANT_METADATA_MOST &&
 		    (level >= zfs_redundant_metadata_most_ditto_level ||
-		    DMU_OT_IS_METADATA(type) || (wp & WP_SPILL))))
+		    DMU_OT_IS_METADATA(type) || (wp & WP_SPILL))))))
```
This didn't even need to be changed.

akashb-22 replied:
my bad, fixed it.

freqlabs commented:
The documentation of this property in `man/man7/zfsprops.7` needs be updated as well.

freqlabs commented on man/man7/zfsprops.7:
```diff
@@ -1444,7 +1444,7 @@ affects only files created afterward; existing files are unaffected.
 This property can also be referred to by its shortened column name,
 .Sy recsize .
-.It Sy redundant_metadata Ns = Ns Sy all Ns | Ns Sy most
+.It Sy redundant_metadata Ns = Ns Sy all Ns | Ns Sy most Ns | Ns Sy none
```
Please bump the date at the top as well.

freqlabs reviewed (approved):
This looks pretty reasonable to me.

ahrens commented on module/zfs/dmu_objset.c:
```diff
@@ -287,7 +287,8 @@ redundant_metadata_changed_cb(void *arg, uint64_t newval)
 	ASSERT(newval == ZFS_REDUNDANT_METADATA_ALL ||
-	    newval == ZFS_REDUNDANT_METADATA_MOST);
+	    newval == ZFS_REDUNDANT_METADATA_MOST ||
+	    newval == ZFS_REDUNDANT_METADATA_NONE);
```
I think that we should implement this in a way that doesn't cause panics on older software (even on debug).  I see a few options for doing this:
1. Use a new key in the properties ZAP object (but ideally, keep the user interface the same).  So when getting the property we would have code something like:
```
potential_key = sprintf("%s$iuv", prop_name); // iuv = Ignore Unknown Values
if (zap_lookup(props_obj, potential_key, &potential_value) == SUCCESS &&
  property_value_is_understood(prop_name, potential_value)) {
    value = potential_value;
} else if (zap_lookup(props_obj, prop_name, &potential_value) == SUCCESS) {
    value = potential_value
} else {
    value = prop_get_default(prop_name);
}
```
This would provide infrastructure to add "unknown but ignorable" values to other properties in the future.
2. Add a new on-disk feature flag, so that once `redundant_metadata=none` is set, the pool can't be brought back to older systems.

akashb-22 replied:
option 1 looks to be preferable, I can try something similar to the above code and get back with a fix.

freqlabs replied:
If the property is unknown how do you know the default value?

freqlabs replied:
Ah, if the property is known but the value is unknown.

Ornias1993 commented:
Question: Does this "duplicate metadata on top of redundancy" also apply to metadata on special vdevs?
If so, this would considerablly increase the viability of using metadata special vdevs!

akashb-22 commented:
yes, this would apply to special vdevs as well.

behlendorf reviewed (commented):
Thanks for updating this, comments inline.

akashb-22 commented on module/zfs/dsl_prop.c:
```diff
@@ -860,6 +882,49 @@ dsl_prop_inherit(const char *dsname, const char *propname,
+		} else {
+			return (1);
+		}
+}
```
updated.

akashb-22 commented on include/sys/dsl_prop.h:
```diff
@@ -106,6 +106,9 @@ int dsl_prop_set_string(const char *dsname, const char *propname,
     zprop_source_t source);
+boolean_t dsl_prop_val_understood(zfs_prop_t prop, uint64_t value);
+void dsl_prop_set_iuv(objset_t *mos, uint64_t zapobj, const char *propname,
+    int intsz, int numints, const void *value, dmu_tx_t *tx);
```
updated.

akashb-22 commented on module/zfs/dsl_prop.c:
```diff
@@ -41,6 +41,7 @@
 
 #define	ZPROP_INHERIT_SUFFIX "$inherit"
 #define	ZPROP_RECVD_SUFFIX "$recvd"
+#define	ZPROP_IUV_SUFFIX "$iuv"
```
Added a comment above dsl_prop_set_iuv().

akashb-22 commented on tests/zfs-tests/include/properties.shlib:
```diff
@@ -27,7 +27,7 @@ typeset -a canmount_prop_vals=('on' 'off' 'noauto')
 typeset -a logbias_prop_vals=('latency' 'throughput')
 typeset -a primarycache_prop_vals=('all' 'none' 'metadata')
-typeset -a redundant_metadata_prop_vals=('all' 'most')
+typeset -a redundant_metadata_prop_vals=('all' 'most' 'none')
```
yes, have added case to test redundant_metadata=all/none in `tests/zfs-tests/tests/functional/inheritance/inherit_001_pos.ksh`

akashb-22 commented on module/zfs/dsl_prop.c:
```diff
@@ -860,6 +882,49 @@ dsl_prop_inherit(const char *dsname, const char *propname,
+	char *iuvstr = kmem_asprintf("%s%s", propname, ZPROP_IUV_SUFFIX);
+	uint64_t val;
+	int iuv_flag = 0;
+	zfs_prop_t prop = zfs_name_to_prop(propname);
```
updated.

akashb-22 commented on module/zfs/dsl_prop.c:
```diff
@@ -860,6 +882,49 @@ dsl_prop_inherit(const char *dsname, const char *propname,
+	zfs_prop_t prop = zfs_name_to_prop(propname);
+	switch (prop) {
+	case ZFS_PROP_REDUNDANT_METADATA:
+		if (*(uint64_t *)value >= ZFS_REDUNDANT_METADATA_NONE)
```
updated.

akashb-22 commented on include/sys/fs/zfs.h:
```diff
@@ -501,7 +501,8 @@ typedef enum {
 	ZFS_REDUNDANT_METADATA_ALL,
-	ZFS_REDUNDANT_METADATA_MOST
+	ZFS_REDUNDANT_METADATA_MOST,
+	ZFS_REDUNDANT_METADATA_NONE
```
I have added both the `ZFS_REDUNDANT_METADATA_SOME` and `ZFS_REDUNDANT_METADATA_NONE` options.
The behavior of some/none values w.r.t OTs:
`ZFS_REDUNDANT_METADATA_NONE` -> critical_metadata (`DMU_OT_OBJSET, DMU_OT_DSL_DIR, DMU_OT_DSL_PROPS, DMU_OT_DSL_DATASET, DMU_OT_MASTER_NODE, DMU_OT_ZVOL_PROP, DMU_OT_POOL_PROPS, DMU_OT_SA, DMU_OT_SA_MASTER_NODE, DMU_OT_SA_ATTR_REGISTRATION, DMU_OT_SA_ATTR_LAYOUTS`)

`ZFS_REDUNDANT_METADATA_SOME` ->`DMU_OT_IS_METADATA(type)`

ahrens replied:
The MOS always uses ZFS_REDUNDANT_METADATA_ALL, so it isn't relevant to talk about the value of DMU_OT_IS_CRITICAL() for MOS object types (DMU_OT_DSL_DATASET, etc).  So the difference between SOME and NONE is primarily that SOME includes DMU_OT_DIRECTORY_CONTENTS whereas NONE does not?

I think what's implemented here is:
* NONE: essentially no extra copies in this dataset (of course the pool-wide metadata is still stored 3x, and some minor bits like SA stuff).  Any single block damage can take out the entire dataset.
* SOME: extra copy of metadata-type objects (including indirect blocks and data blocks) (e.g. dnodes, directories).  Any single block damage can take out at most one user file (or zvol).
* MOST: additionally, extra copy of L2 and above indirect blocks of user-type objects (e.g. files, zvols).  Any single block damage can take out at most 1024 logically-contiguous blocks of user data (e.g. 8MB, with recordsize=8k).
* ALL (default): additionally, extra copy of L1 indirect blocks of user-type objects.  Any single block damaged can take out at most 1 block of user data (e.g. 128K, with recordsize=128k).

It seems to me that distinguishing "critical" object types might not be necessary, if we are OK with NONE not having extra copies of the SA and other small bits of metadata.  It's true that they are critical for operating on the dataset, but you could say the same thing about the single L5 indirect block of a zvol, or the block of dnodes that contains the root directory.  So with NONE, any corruption in the dataset has a chance of losing the whole dataset, regardless of the "critical" object types.

I think it might be hard for admins to chose among these 4 options.  Maybe documenting (in the manpage) the "any single block damage can take out..." for each of the options, would help them understand the trade-off.  We currently do that for ALL and MOST, although in a slightly hand-wavey way, and also it says 100 where it should say 1000 (due to the long-ago indirect block size increase from 16KB -> 128KB).

akashb-22 commented:
A few of the tests are failing in CI due to
```
/usr/share/zfs/zfs-tests/tests/functional/cli_root/zpool_import/zpool_import_features_001_pos.ksh[60]: log_must[70]: log_pos: line 265: 781205: Abort(coredump)
module/zfs/dmu.c:2049:7: runtime error: index 195 out of bounds for type 'dmu_object_type_info_t [54]'
    #0 0x7f157c809b98 in dmu_write_policy module/zfs/dmu.c:2049
    #1 0x7f157c7b6fc2 in dbuf_write module/zfs/dbuf.c:5028
    #2 0x7f157c7e23b2 in dbuf_sync_leaf module/zfs/dbuf.c:4498
    #3 0x7f157c7e4964 in dbuf_sync_list module/zfs/dbuf.c:4537
    #4 0x7f157c8e3085 in dnode_sync module/zfs/dnode_sync.c:850
    #5 0x7f157c81a1c4 in dmu_objset_sync_dnodes module/zfs/dmu_objset.c:1564
    #6 0x7f157c81a5a6 in sync_dnodes_task module/zfs/dmu_objset.c:1634
    #7 0x7f157c62f1d6 in taskq_thread lib/libzpool/taskq.c:240
    #8 0x7f157c628cb3 in zk_thread_wrapper lib/libzpool/kernel.c:89
    #9 0x7f157ae1d608 in start_thread /build/glibc-SzIz7B/glibc-2.31/nptl/pthread_create.c:477
    #10 0x7f157ad40132 in __clone (/lib/x86_64-linux-gnu/libc.so.6+0x11f132)

ERROR: zhack -d /var/tmp/dev_import-test feature enable testpool1 com.test:xxx_unsup0 exited 262
NOTE: Performing test-fail callback (/usr/share/zfs/zfs-tests/callbacks/zfs_dbgmsg.ksh)
```
However, I'm not able to reproduce the above issue in my environment setup. Can anyone help me?

ahrens commented on module/zfs/dsl_prop.c:
```diff
@@ -647,6 +673,46 @@ dsl_prop_changed_notify(dsl_pool_t *dp, uint64_t ddobj,
+    int intsz, int numints, const void *value, dmu_tx_t *tx)
+{
+	char *iuvstr = kmem_asprintf("%s%s", propname, ZPROP_IUV_SUFFIX);
+	int iuv_flag = 0;
```
it looks like this could be more simply, `boolean_t iuv = B_FALSE`

akashb-22 replied:
updated.

ahrens commented on module/zfs/dsl_prop.c:
```diff
@@ -647,6 +673,46 @@ dsl_prop_changed_notify(dsl_pool_t *dp, uint64_t ddobj,
+		VERIFY0(zap_update(mos, zapobj, propname, intsz, numints,
+		    &val, tx));
+	} else if (0 == zap_contains(mos, zapobj, iuvstr)) {
+		zap_remove(mos, zapobj, iuvstr, tx);
```
there's no need to check zap_contains() first.

```
/*
 * Remove the specified attribute.
 *
 * If the specified attribute does not exist, the call will fail and
 * return ENOENT.
 */
int zap_remove(objset_t *ds, uint64_t zapobj, const char *name, dmu_tx_t *tx);
```

akashb-22 replied:
updated.

ahrens commented on module/zfs/dsl_prop.c:
```diff
@@ -69,6 +70,16 @@ dodefault(zfs_prop_t prop, int intsz, int numints, void *buf)
+	if (zfs_prop_get_type(prop) == PROP_TYPE_INDEX)
+		return (!zfs_prop_index_to_string(prop, value, &str));
+
+	return (-1);
```
We typically use return value of `0` for success, and positive errno's for errors.  so I would use `return (EINVAL)` here.  Or it might be simpler to return `boolean_t` since really we are asking, "is this known?  yes/no"

akashb-22 replied:
Here `dsl_prop_known_index` returns 1 if the property index is known, returns 0 if the property index is unknown, and returns -1 if it is not an index-type property. @behlendorf has mentioned [here](https://github.com/openzfs/zfs/pull/13680#discussion_r966268129) this may be useful in the future.

behlendorf commented:
There are old and new object types which we need to be careful to check for in this macro.  You'd want to do something like  `DMU_OT_IS_METADATA` does and check `DMU_OT_NEWTYPE`.  This is what's causing the test failure.

```c
#define DMU_OT_IS_METADATA(ot) (((ot) & DMU_OT_NEWTYPE) ? \
        (((ot) & DMU_OT_METADATA) != 0) : \
        DMU_OT_IS_METADATA_IMPL(ot))
```
  
That said, to make sure we're targeting only the highest impact types I dumped the usage by-object-type, with `zdb -bbb`, from a couple production Lustre MDTs.  As expected the biggest consumers were the `DMU_OT_DNODE`, `DMU_OT_DIRECTORY_CONTENTS`, and `DMU_OT_SA` objects.  Everything else tagged as metadata was functionally insignificant.  Making redundant copies of those costs of almost nothing.  Given that, we can simplify the "some" policy to all metadata except these three types.  i.e.

```c
#define DMU_OT_IS_CRITICAL(ot) (((ot) & DMU_OT_NEWTYPE) ? \
        (((ot) & DMU_OT_METADATA) != 0) : \
        (DMU_OT_IS_METADATA_IMPL(ot) && \
        (ot != DMU_OT_DNODE) && \
        (ot != DMU_OT_DIRECTORY_CONTENTS) && \
        (ot != DMU_OT_SA))))
```

With Data-on-MDS enabled, only ~13% of the total capacity is consumed by ZFS-tagged metadata on a MDT.  So just to reiterate, the big win from the "some" option isn't capacity but create/unlink performance.

```
Blocks  LSIZE   PSIZE   ASIZE     avg    comp   %Total  Type
 58.0M  3.38T   3.38T   3.38T   59.6K    1.00    86.88  ZFS plain file
 48.6M  70.5G   70.5G    388G      8K    1.00     9.76  System attributes
 10.8M   561G   45.6G   91.3G   8.46K   12.30     2.29  ZFS directory
 3.85M  62.2G   19.5G   38.9G   10.1K    3.19     0.98  DMU dnode
 18.0K  2.11G   1.24G   3.73G    212K    1.70     0.09  SPA space map
                                                 <0.01  <everything else>  
```

ahrens commented:
@behlendorf your macro can be implemented in terms of DMU_OT_IS_METADATA:
```c
#define DMU_OT_IS_CRITICAL(ot) \
        (DMU_OT_IS_METADATA(ot) && \
        (ot) != DMU_OT_DNODE && \
        (ot) != DMU_OT_DIRECTORY_CONTENTS && \
        (ot) != DMU_OT_SA)
```
Although as I mentioned in my other comment, maybe we don't really care about keeping an extra copy of these "critical" objects, since there are still several unduplicated blocks that could take out access to the entire dataset (e.g. the first block of dnodes, the root directory contents, etc).

akashb-22 commented:
@ahrens  @behlendorf 
If any corruption in the dataset has a chance of losing the whole dataset, regardless of the "critical" object types then it looks like distinguishing "critical" object types might not be necessary. As MOS anyway uses ZFS_REDUNDANT_METADATA_ALL, at least pool-wide metadata is still stored redundantly. So, I have updated the "none" policy for no extra metadata copies and the "some" policy to have extra copies for all metadata except these three types.
i.e:
```#define DMU_OT_IS_CRITICAL(ot) \
        (DMU_OT_IS_METADATA(ot) && \
        (ot) != DMU_OT_DNODE && \
        (ot) != DMU_OT_DIRECTORY_CONTENTS && \
        (ot) != DMU_OT_SA)
```
Also, I have modified the man page accordingly.

akashb-22 commented:
zfs-tests-functional test seems to be failing due to some other issues like many other CI jobs which are failing due to timeout (Error: The action has timed out.).

behlendorf commented on man/man7/zfsprops.7:
```diff
@@ -1495,6 +1495,16 @@ of user data can be lost if a single on-disk block is corrupt.
 .Pp
+When set to
+.Sy some ,
+ZFS stores an extra copy of few types of critical metadata.
```
```suggestion
ZFS stores an extra copy of only critical metadata.
```

akashb-22 replied:
done.

behlendorf commented on man/man7/zfsprops.7:
```diff
@@ -1495,6 +1495,16 @@ of user data can be lost if a single on-disk block is corrupt.
+.Pp
+When set to
+.Sy none ,
+ZFS does not store any copies of metadata redundantly through this property.
```
```suggestion
ZFS does not store any copies of metadata redundantly.
```

akashb-22 replied:
done.

behlendorf commented on man/man7/zfsprops.7:
```diff
@@ -1495,6 +1495,16 @@ of user data can be lost if a single on-disk block is corrupt.
+When set to
+.Sy none ,
+ZFS does not store any copies of metadata redundantly through this property.
+If a single on-disk block is corrupt, entire dataset can be lost.
```
```suggestion
If a single on-disk block is corrupt, an entire dataset can be lost.
```

akashb-22 replied:
done

behlendorf commented on man/man7/zfsprops.7:
```diff
@@ -1495,6 +1495,16 @@ of user data can be lost if a single on-disk block is corrupt.
+When set to
+.Sy some ,
+ZFS stores an extra copy of few types of critical metadata.
+If a single on-disk block is corrupt, at worst a single user file can be lost.
```
I'd add, "This can improve file create performance since less metadata needs to be written."

akashb-22 replied:
done.

behlendorf commented on module/zfs/dmu.c:
```diff
@@ -1993,6 +1993,9 @@ dmu_write_policy(objset_t *os, dnode_t *dn, int level, int wp, zio_prop_t *zp)
 		if (os->os_redundant_metadata == ZFS_REDUNDANT_METADATA_ALL ||
+		    (os->os_redundant_metadata ==
+		    ZFS_REDUNDANT_METADATA_SOME &&
+		    DMU_OT_IS_CRITICAL(type)) ||
```
It's functionally equivalent but let's convert this to a switch statement to make the policy more readable.  Something like this:

```c
                switch (os->os_redundant_metadata) {
                case ZFS_REDUNDANT_METADATA_ALL:
                        copies++;
                        break;
                case ZFS_REDUNDANT_METADATA_MOST:
                        if (level >= zfs_redundant_metadata_most_ditto_level ||
                            DMU_OT_IS_METADATA(type) || (wp & WP_SPILL))
                                copies++;
                        break;
                case ZFS_REDUNDANT_METADATA_SOME:
                        if (DMU_OT_IS_CRITICAL(type))
                                copies++;
                        break;
                case ZFS_REDUNDANT_METADATA_NONE:
                        break;
                }
```

akashb-22 replied:
yep, changed it. Thanks for the suggestion.

behlendorf commented on module/zfs/dsl_prop.c:
```diff
@@ -647,6 +673,46 @@ dsl_prop_changed_notify(dsl_pool_t *dp, uint64_t ddobj,
+ * versions and the default property value from the existing prop key is
+ * used.
+ */
+
```
no need for the extra newline here.

akashb-22 replied:
done.

CyberCr33p commented:
@akashb-22 do you have any benchmarks with redundant_metadata=most ?
